### PR TITLE
server,client: Add_split_and_scatter_region 

### LINKED ISSUE #4095: Enhance Split Table Stability and Usability
## Description

In the [high concurrency practice of TiDB](https://docs.pingcap.com/tidb/stable/high-concurrency-best-practices), split table is a common operation for users. It can help users solve hotspot issues in advance and prevent sudden traffic from being unable to handle it.

## Document Collection

- Proposal doc: ${proposal doc}
- Weekly report: ${weekly report link}

## Talent Challenge Program information

- Mentor of this issue: @Yisaer 
- Recommended skills:  Go
- Estimated Workloads: 2XL

## Milestones and action items

Milestone 1:  Enhance the `split table` stability

- [ ] dive to the problem https://github.com/pingcap/tidb/issues/24295
- [ ] Let PD check regions before splitting regions https://github.com/tikv/pd/issues/2733
- [ ] All downstream use new API.

Milestone 2: Enhance the `split table` usability
refer: https://github.com/wentaojin/tidba, we can extend the `split table` syntax.

- [ ] TBD
- [ ] TBD

### PR DESCRIPTION
**Pull request by qidi1**
### What problem does this PR solve?

Issue Number: ref #4095

The original region split and scatter are handled by tikv and pd respectively, in order to improve efficiency, the split is also changed to be handled and controlled by pd.

### What is changed and how it works?

Add a method splitAndScatterRegions in grpc-service
let pd handle split and scatter regions
Code changes based on this [work](https://github.com/hzh0425/pd/tree/add_split_and_scatter_region)
Thanks to [hzh0435](https://github.com/hzh0425)

### Check List

Tests

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Code changes

- [ ] Has configuration change
- [ ] Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- [ ] Has persistent data change

Side effects

- [ ] Possible performance regression
- [ ] Increased code complexity
- [ ] Breaking backward compatibility

Related changes

- [ ] PR to update [`proposal`](https://github.com/pingcap/tidb/pull/33664/files)

### Release note
```release-note
None
```

### CODE REVIEW

**Comment by qidi1** [comment]
I don't know where to test the client package code. Please tell me if anyone knows

File: server/grpc_service.go [hunk 1]
```diff
@@ -1487,8 +1487,56 @@ func (s *GrpcServer) SplitRegions(ctx context.Context, request *pdpb.SplitRegion
...
+	return &pdpb.SplitAndScatterRegionsResponse{
+		Header:                    s.header(),
+		RegionsId:                 newRegionIDs,
+		SplitFinishedPercentage:   uint64(splitFinishedPercentage),
+		ScatterFinishedPercentage: uint64(scatterFinishedPercentage),
+	}, nil
+}
+
+// addScatterRegionsOperators add operators to scatter regions and return the processed percentage and error
+func (s *GrpcServer) addScatterRegionsOperators(cluster *cluster.RaftCluster, regionsID []uint64, group string, retryLimit int) (int, error) {
```
Comment (rleungx) [inline server/grpc_service.go:1518]:
```suggestion
func scatterRegions(cluster *cluster.RaftCluster, regionsID []uint64, group string, retryLimit int) (int, error) {
```

Comment (qidi1) [reply server/grpc_service.go]: I have changed

File: server/grpc_service.go [hunk 2]
```diff
@@ -1487,8 +1487,56 @@ func (s *GrpcServer) SplitRegions(ctx context.Context, request *pdpb.SplitRegion
 }
 
 // SplitAndScatterRegions split regions by the given split keys, and scatter regions
-func (s *GrpcServer) SplitAndScatterRegions(_ context.Context, _ *pdpb.SplitAndScatterRegionsRequest) (*pdpb.SplitAndScatterRegionsResponse, error) {
-	panic("unimplemented")
+func (s *GrpcServer) SplitAndScatterRegions(ctx context.Context, request *pdpb.SplitAndScatterRegionsRequest) (*pdpb.SplitAndScatterRegionsResponse, error) {
+	forwardedHost := getForwardedHost(ctx)
```
[the remark this answers is not in the document]
Comment (qidi1) [reply server/grpc_service.go]: I have changed

File: server/grpc_service.go [hunk 3]
```diff
@@ -1487,8 +1487,56 @@ func (s *GrpcServer) SplitRegions(ctx context.Context, request *pdpb.SplitRegion
...
+		}
+		ctx = grpcutil.ResetForwardContext(ctx)
+		return pdpb.NewPDClient(client).SplitAndScatterRegions(ctx, request)
+	}
+	if err := s.validateRequest(request.GetHeader()); err != nil {
+		return nil, err
+	}
+	rc := s.GetRaftCluster()
+	splitFinishedPercentage, newRegionIDs := rc.GetRegionSplitter().SplitRegions(ctx, request.GetSplitKeys(), int(request.GetRetryLimit()))
+	scatterFinishedPercentage, err := s.addScatterRegionsOperators(rc, newRegionIDs, request.GetGroup(), int(request.GetRetryLimit()))
```
[the remark this answers is not in the document]
Comment (qidi1) [reply server/grpc_service.go]: yes,

> The scatterFinishedPercentage only reflects the progress of the processed split regions, right?

[the remark this answers is not in the document]
Comment (rleungx) [reply server/grpc_service.go]: So when there are regions that are not split successfully, the whole request needs to retry no matter what the scatter status is, right?

[the remark this answers is not in the document]
Comment (qidi1) [reply server/grpc_service.go]: If you want all the splits of the specified key range to be successful and to be scattered. Must retry the whole

File: client/client.go [hunk 4]
```diff
@@ -1693,6 +1695,36 @@ func (c *client) ScatterRegions(ctx context.Context, regionsID []uint64, opts ..
...
+	}
+	req := &pdpb.SplitAndScatterRegionsRequest{
+		Header:     c.requestHeader(),
+		SplitKeys:  splitKeys,
+		Group:      options.group,
+		RetryLimit: options.retryLimit,
+	}
+
+	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
+	resp, err := c.getClient().SplitAndScatterRegions(ctx, req)
```
Comment (rleungx) [inline client/client.go:1719]: How about:
```suggestion
	retrun c.getClient().SplitAndScatterRegions(ctx, req)
```

File: server/grpc_service.go [hunk 5]
```diff
@@ -1633,8 +1617,52 @@ func (s *GrpcServer) SplitRegions(ctx context.Context, request *pdpb.SplitRegion
...
+	if err != nil {
+		return 0, err
+	}
+	for _, op := range ops {
+		if ok := cluster.GetOperatorController().AddOperator(op); !ok {
+			failures[op.RegionID()] = fmt.Errorf("region %v failed to add operator", op.RegionID())
+		}
+	}
+	percentage := 100
+	if len(failures) > 0 {
```
Comment (rleungx) [inline server/grpc_service.go:1655]: How about using a variable for `len(failures)`?

File: server/grpc_service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rleungx) [reply server/grpc_service.go]: prefer to add a comment

File: server/grpc_service.go (hunk 5 above)
Comment (qidi1) [reply server/grpc_service.go:1509]: pardon? I see many palce used like this for example handle
image

File: client/client.go (hunk 4 above)
Comment (qidi1) [reply client/client.go]: I have changed

File: server/grpc_service.go (hunk 5 above)
Comment (rleungx) [reply server/grpc_service.go:1655]: maybe we can abstract a function to calculate the percentage?

Comment (qidi1) [reply server/grpc_service.go:1655]: there is no raft cluster in scatterRegion. If we want abstract a function to calculate.,we have to change cluster to raftcluster in ReionScatterer
image
but I found

Comment (rleungx) [reply server/grpc_service.go:1655]: Got it. We can leave it for now.

File: server/grpc_service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (qidi1) [reply server/grpc_service.go]: I have add a comment

**Review by rleungx — APPROVED** [review]

**Comment by bufferflies** [comment]
It will be better to rename the title of this pr,  how about: "server: implemente `SplitAndScatterRegions` and remove the code from the client "

**Review by bufferflies — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**